web/assets/js/components/gameboard.ts

```diff
@@ -144,22 +153,87 @@ const Gameboard = class extends Component {
       this.call('stopTimer', [true], 'gameplayers', 'rival');
       this.call('disappear', '', 'gamecontrols');
 
-      return; // action
+      return;
     }
 
-    //game turn update
+    // Game turn update (player timer and action balancing)
     if (this.color === this.chess.turn()) {
-      this.call('stopTimer', [init], 'gameplayers', 'rival');
-      this.call('startTimer', [init], 'gameplayers', 'me');
+      // My turn to play
+      // If first move, 30s rule - no timer
       this.allowedToMove = true;
+      this.call('stopTimer', [init], 'gameplayers', 'rival');
+
+      //TODO: chech if this.getMoveNumber() return 0 or 1 for before first move.
+      if (this.getMoveNumber() > 1) {
+        const dateStartGame = gameState.time?.started_at;
+
+        if (!dateStartGame) {
+          throw new Error('No started_at game time');
+          return;
+        }
+        // Get the time of creation & current time
+        // Get the diff and use it to get the updated / sync timeout by substract by 30s
+        const startDate = new Date(dateStartGame);
+        const currentDate = new Date();
+        const diffDate = startDate.getTime() - currentDate.getTime();
+
+        const exitforTimeout = async () => {
+          try {
+            // Claim timeout. If no error, timeout succeeded
+            await actions.claimTimeout(this.gameId);
+            this.call(this.engine(false, 'timeout'));
+          } catch (e) {
+            this.call(
+              'appear',
+              ['Invalid claim timeout request', 'error'],
+              'toast'
+            );
+            this.call(this.engine(false, 'timeout')); //TODO: create a fail "exit" gameover in the engine
+            // Timeout request is invalid
+            // for the user (I assume fire event to end game)
+          }
+        };
+
+        const timer = 30 * 1000 - diffDate;
+
+        if (timer <= 0) {
+          // if date too old (> 30sec )
+          await exitforTimeout();
+          throw new Error('Game party started too long time ago');
+        }
+
+        this.initMoveTimer = setTimeout(async () => {
+          await exitforTimeout();
+        }, timer); //30sec first move
+      } else {
+        this.call('startTimer', [init], 'gameplayers', 'me');
```

Comment:
So all the logic for the normal case is in startTimer, right?

Reply:
Not really. This function is called in the engine function that is used as player balancer. The engine function checks for each turn if there is a game over and if not, give or not the right to play and start / stop the timers accordingly. 

This engine function is called after the current player has moved a piece and receiving a move from the server. It basically wait this rival move promise to be completed to let the user play. During this time, the user cannot select pieces and only see the rival timer decreasing. 
It's the same for both players but not in the same time. 

This function you mention is called to update the timers in the 30s rule exception.

Reply:
@thehowl